### PR TITLE
Load deamon.json configuration

### PR DESCRIPTION
**Pull request by Sventimir**
This PR creates the possibility to load genesis ledger and protocol constants from `daemon.json` file that is used by OCaml nodes. Not everything is currently read from that file; only the values that we might want to use at the moment. The format can be extended to encompass all possible options understood by OCaml nodes if required.

### CODE REVIEW

File: node/src/daemon_json/json_genesis.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    pub fn slots_per_sub_window(&self) -> UnsignedExtendedUInt32StableV1 {
+        UnsignedExtendedUInt32StableV1(Number(self.slots_per_sub_window))
+    }
+
+    pub fn grace_period_slots(&self) -> UnsignedExtendedUInt32StableV1 {
+        UnsignedExtendedUInt32StableV1(Number(self.grace_period_slots))
+    }
+
+    pub fn genesis_state_timestamp(&self) -> anyhow::Result<BlockTimeTimeStableV1> {
```
Comment (akoptelov) [inline node/src/daemon_json/json_genesis.rs:43]: I wouldn't use `anyhow` based errors in library crates.

File: node/src/daemon_json/json_genesis.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                BlockTimeTimeStableV1(UnsignedExtendedUInt64Int64ForVersionTagsStableV1(Number(
+                    dt.timestamp() as u64,
+                )))
+            })
+            .map_err(|e| anyhow::anyhow!(e))
+    }
+
+    pub fn to_genesis_constants_checked(
+        &self,
+    ) -> anyhow::Result<MinaBaseProtocolConstantsCheckedValueStableV1> {
```
Comment (akoptelov) [inline node/src/daemon_json/json_genesis.rs:55]: See above.

**Review by akoptelov — CHANGES_REQUESTED** [review]
I think there should be a special error type instead of `anyhow`-based one.

And also there are some warnings from clippy that might need fixing.

**Comment by Sventimir** [comment]
> And what about specifying daemon.json from node's command line?
@akoptelov, I fear this won't work just by itself. For a custom network you either need to operate in seed node (which I believe Openmina cannot do at the moment) or specify a custom peer list (which I don't know if is possible). I can implement the option switch, but it won;t be very useful unless we can also specify a custom peer list.

**Comment by akoptelov** [comment]
@Sventimir different set of seeds is the least of the issues:
```
cargo run --release -p cli node --peers \
  /ip4/34.48.73.58/tcp/10003/p2p/12D3KooWAdgYL6hv18M3iDBdaK1dRygPivSfAfBNDzie6YqydVbs \
  /ip4/35.245.82.250/tcp/10003/p2p/12D3KooWLjs54xHzVmMmGYb7W5RVibqbwD1co7M2ZMfPgPm7iAag \
  /ip4/35.245.106.190/tcp/10003/p2p/12D3KooWEiGVAFC7curXWXiGZyMWnZK9h8BKr88U8D5PKV3dXciv
```
Another thing is to specify a proper chain id. We don't have that option anymore, the idea is to derive it from constants/genesis, but we can revert that at least temporarily.

[0001-chain_id.patch.gz](https://github.com/openmina/openmina/files/15127784/0001-chain_id.patch.gz)

**Comment by Sventimir** [comment]
@akoptelov, chain id can and indeed should be derrived from genesis parameters. Okay, I didn't realize we can pass a peer list. In that case I'll add the option to load a config at startup.

File: node/src/daemon_json/json_genesis.rs (hunk 1 above)
Comment (Sventimir) [reply node/src/daemon_json/json_genesis.rs]: Done in d3d4829f.

File: node/src/daemon_json/json_genesis.rs (hunk 2 above)
Comment (Sventimir) [reply node/src/daemon_json/json_genesis.rs]: Done in: d3d4829f.

**Comment by Sventimir** [comment]
I added a (mandatory) option to specify a config file. Unfortuantely,. when supplied wityh the berkeley JSON config, the node crashes:
```
thread 'main' panicked at /home/sven/work/openmina/node/src/transition_frontier/sync/transition_frontier_sync_effects.rs:50:62:
called `Result::unwrap()` on an `Err` value: "Tried to copy from non-existing snarked ledger with hash: jwkqwgAC6MXgfiZmynHRqXV6PGbMbLwFCx56Y2rt5vwdumf6ofp"
```

**Comment by tizoc** [comment]
The issue for that bug is here https://github.com/openmina/openmina/issues/331, if you launch the node again it may work this time.

In theory it should have been fixed by https://github.com/openmina/openmina/pull/342 but it seems thats not the case

**Comment by tizoc** [comment]
@Sventimir actually, your seems to be different, the problem is that the genesis ledger is not there

File: node/src/transition_frontier/genesis/transition_frontier_genesis_config.rs [hunk 3]
```diff
@@ -130,6 +135,44 @@ impl GenesisConfig {
...
+                    .ok_or(anyhow::anyhow!("No accounts in config"))?;
+                let accounts: Vec<Account> = account_list
+                    .iter()
+                    .map(daemon_json::Account::to_account)
+                    .collect::<Result<Vec<Account>, AccountConfigError>>()
+                    .map_err(anyhow::Error::from)?;
+                let (mut mask, total_currency) = Self::build_ledger_from_accounts(accounts);
+                let result = GenesisConfigLoaded {
+                    constants,
+                    ledger_hash: ledger_hash(&mut mask),
```
Comment (tizoc) [inline node/src/transition_frontier/genesis/transition_frontier_genesis_config.rs:170]: @Sventimir which hash are you getting here when using the daemon.json file? issue could be that you are not inserting the winner account (check the ledger tool and how it generates it)

https://github.com/openmina/openmina/blob/develop/tools/ledger-tool/src/main.rs#L128-L138

Comment (Sventimir) [reply node/src/transition_frontier/genesis/transition_frontier_genesis_config.rs:185]: I added the genesis winner account in https://github.com/openmina/openmina/pull/369/commits/629b87e8eaa99414634b19ab9311ba837b46b4cc, but it does not seem to help much. I still get the same error.

Comment (tizoc) [reply node/src/transition_frontier/genesis/transition_frontier_genesis_config.rs:185]: @Sventimir try adding the genesis account first, then the rest

Comment (Sventimir) [reply node/src/transition_frontier/genesis/transition_frontier_genesis_config.rs:185]: Did that, didn't help either. The ledger hash I get is different than expected: `jxA9yKKm77xdUbCBp59bdaTVRD9MqzX8HorKyg5sghFQeckUxnL`.

**Comment by Sventimir** [comment]
Rebased and fixed problems with loading network config from file. The node does sync again.

File: tools/ledger-tool/src/main.rs [hunk 4]
```diff
@@ -148,6 +148,7 @@ fn main() -> anyhow::Result<()> {
     }
 
     for account in &accounts {
+        println!("{account:?}");
```
Comment (tizoc) [inline tools/ledger-tool/src/main.rs:151]: leftover print

**Review by tizoc — APPROVED** [review]
@Sventimir LGTM, please squash and cleanup the commits in any way you think makes sense and merge once you are happy with it

File: cli/src/commands/node/mod.rs [hunk 5]
```diff
@@ -158,9 +166,17 @@ impl Node {
             )
         });
 
+        let config_file = self.config.map(File::open).transpose()?;
```
Comment (akoptelov) [inline cli/src/commands/node/mod.rs:169]: If incorrect file will be specified, something like `[ERROR] No such file or directory (os error 2)`, with no clue which file is not found. I would use `anyhow::Context` here, but for some reason we're using boxed Error in `cli`, so probably an `error` log entry is worth adding here.

File: cli/src/commands/node/mod.rs [hunk 6]
```diff
@@ -158,9 +166,17 @@ impl Node {
...
+        let conf: Option<DaemonJson> = config_file.map(serde_json::from_reader).transpose()?;
```
Comment (akoptelov) [inline cli/src/commands/node/mod.rs:170]: The same as above.

**Review by akoptelov — CHANGES_REQUESTED** [review]
LGTM, with two minor fixes to add.

File: cli/src/commands/node/mod.rs (hunk 5 above)
Comment (Sventimir) [reply cli/src/commands/node/mod.rs:169]: Does this look better:
```
2024-05-09T09:19:20.509149Z ERROR openmina::commands::node: time: "1715246360507319345", kind: "ConfigFileError", summary: "failed to open config file", error: "Os { code: 2, kind: NotFound, message: \"No such file or directory\" }"
[ERROR] No such file or directory (os error 2)
```
?

File: cli/src/commands/node/mod.rs (hunk 6 above)
Comment (Sventimir) [reply cli/src/commands/node/mod.rs:170]: Like this:
```
2024-05-09T09:22:07.440473Z ERROR openmina::commands::node: time: "1715246527440454455", kind: "ConfigFileError", summary: "failed to parse config file", error: "Error(\"expected value\", line: 1, column: 1)"
[ERROR] expected value at line 1 column 1
```
?

**Review by akoptelov — APPROVED** [review]

**Pull request merged**